trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracestate
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+const (
+	keyMaxSize       = 256
+	valueMaxSize     = 256
+	maxKeyValuePairs = 32
+)
+
+const (
+	keyWithoutVendorFormat = `[a-z][_0-9a-z\-\*\/]{0,255}`
+	keyWithVendorFormat    = `[a-z][_0-9a-z\-\*\/]{0,240}@[a-z][_0-9a-z\-\*\/]{0,13}`
+	keyFormat              = `(` + keyWithoutVendorFormat + `)|(` + keyWithVendorFormat + `)`
+	valueFormat            = `[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]`
+)
+
+var keyValidationRegExp = regexp.MustCompile(`^(` + keyFormat + `)$`)
+var valueValidationRegExp = regexp.MustCompile(`^(` + valueFormat + `)$`)
+
+// Tracestate represent tracing-system specific context in a list of key-value pairs. Tracestate allows different
+// vendors propagate additional information and inter-operate with their legacy Id formats.
+type Tracestate struct {
+	tracestateEntries []*TracestateEntry
+}
+
+// TracestateEntry represent one key-value pair in a list of key-value pair of Tracestate.
+// Key is an opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type TracestateEntry struct {
+	key   string
+	value string
+}
+
+// Key returns the key of TracestateEntry
+func (te *TracestateEntry) Key() string {
+	return te.key
+}
+
+// Value returns the value of TracestateEntry
+func (te *TracestateEntry) Value() string {
+	return te.value
+}
+
+// Get retrieves value for a given key from Tracestate ts.
+// If the key is not found then false is returned with the value "" is returned.
+// If the key is found then true is returned with its value.
+func (ts *Tracestate) Get(key string) (string, bool) {
+	if len(ts.tracestateEntries) == 0 {
+		return "", false
+	}
+	for _, entry := range ts.tracestateEntries {
```

Comment:
Why isn't we keep all the entries in a map. This is an O(n) lookup.

Reply:
We need to maintain ordered list of key-value pairs. Yes, map would be efficient in searching. We would need both. we can keep map[string]int for storing key-arrayIndex to improve the search. But the number of key-value pair is not expected to be high (based on comment from Java).
I can certainly make the change to add the key/arrayIndex for efficient search.

Reply:
I indeed don't expect this number to be high... ask yourself... will we expect more than a couple of different tracing systems collaborating in a request be a common use case? So I don't think we need additional code for more efficient search at this point (premature optimization).